### ISSUE TITLE
RFC Sustainability of Mitogen project

### ISSUE THREAD

**Issue by moreati** [mitogen-hq/mitogen#773]
The Mitogen project is currently hosted on accounts controlled by David, the project's creator. We have not been in contact with David for some time. To continue maintaining Mitogen we need to either get in contact with David, and agree a plan. Or if we are unable to reach him we might have to fork the project under a new organisation.

The trouble spots are
- Inability to add or remove CI checks and hence merge PRs
- Inability to make new releases to PyPI
- Inability to manage roles of maintainers and increase the bus factor

ETA: 13 Jan 2020 Added summary following asking for comments from wider community.
Edit 22 Jan 2021: Docker Hub rate limiting moved to #791

**Comment by moreati** [comment]
Related questions (that could be worth promoting to distinct issues)
- Half the tests run on travis-ci.org, which is "shutting down in several weeks".
  - What should we do?
- Half the tests run on Azure Pipelines
  - Are we costing @dw money by running CI tests?
- Much of the project infrastructure (e.g. CI, PyPI releases) run under @dw's credentials but he is less involved in this project.
  - How should we continue?
  - Is the project sustainable?

**Comment by moreati** [comment]
@s1113950 it might be worth catching up in e.g. IRC, Zoom.  I'm in irc://freenode.org/mitogen most evenings, UTC.

**Comment by s1113950** [comment]
@moreati good idea! I'm in PST (UTC-08). Here are my initial thoughts on your questions. If you'd like to discuss things on something like IRC further respond here and we can figure something out :)

> Who's account?

I'm not sure on this, I would assume it's @dw 's account.

> Paid for by who?

Is it possible to get sponsored by an org that uses Mitogen perhaps?

> Can we embed a secret/token in the repo, given constraints?

We get more requests if we do this I believe; we could also change the account to a dedicated Mitogen account and then for sure we'd be under the request limit.

> Switch to an alternative container registry

I'm not sure on limits of other container registries, but JFrog's Free tier has 2GB of storage for free (so depending on required image size we might be able to use this). There's also CircleCI, see caching question below.

> Perhaps with caching, or pinning?

If we switch to CircleCI they have caching support, and a generous open source tier as well: https://circleci.com/open-source/ .

> What should we do?

@dw is the only one that can change this afaik, and I've contacted him 2 weeks ago and have not received a reply. My last response from him was in May; I'm not sure how active (if at all) he is with Mitogen anymore (haven't seen him reply to any Mitogen Github ticket in > 6 months).
Since he hasn't responded, the only options I can think of at this point are to either make a new Mitogen org somewhere and continue the project there (github.com/mitogen is taken) or continue it off of a fork. In either case, we'd push a change to the README of Mitogen (before travis.org dies and we can't merge PRs anymore to `master`) that notifies people of the new change.

> Are we costing @dw money by running CI tests?

He hasn't responded to me in awhile so I'm not sure if the tests cost $$$; the tests themselves run off of the free tiers of both Azure and TravisCI it looks like so my guess is there's no cost associated.

> How should we continue?

Answer similar to the `What should we do` question. This needs to change as well if the project is to continue. I can't push to `PyPi` and it's quite frustrating having like 1/2 "admin power" on a project that you're maintaining.

> Is the project sustainable?

This is the big question I've been debating recently. I think it's a good project (in its current state it runs Ansible tests about 5X faster than vanilla Ansible which is pretty impressive imo), but question its future given that Operon was announced. Also, I'm getting less and less motivated to keep working on it because I'm supposed to be maintaining it yet can't push to PyPi, change `Settings` on the project for when things like TravisCI EOL happen, and don't have access to the Docker account. There's also very few ppl actively making PRs for a project this size.

In the immediate term: something has to change for Mitogen to continue to be maintained, because once travisci.org is EOL, I can't merge PRs anymore. I could start working off of another branch like `develop` that would have no test requirements, but I'm tired of having to do workarounds in order to maintain a project that I don't actively use anymore 😞

**Comment by moreati** [comment]
Sorry, I assumed you had a lot more access to the Github project than you do. e.g. you could configure PR checks such as Travis.

I agree, without help from David we can't continue Mitogen under this Github account. We would need to move it, possibly fork it under a new name. I _really_ want to avoid that - particularly renaming it - if possible. Mitogen is David's creation, and it should stay that way even if he has moved on to other things.

Beyond any organisational issues, I'm not sure a successor to mitogen would be sustainable. I fundamentally don't understand the fork/fd/signal/socket contortions that Mitogen does. I don't have the time or knowledge to be a core maintainer.

> I'm tired of having to do workarounds in order to maintain a project that I don't actively use anymore

You shouldn't have to put effort into a tool you're not benefitting from.

Perhaps it's best to get 0.3.0 out the door, but make it clear there won't be any more in the foreseeable furture. Or leave the release at `0.3.0-beta1`. Let Mitogen slumber.

**Comment by moreati** [comment]
@s1113950 to clear any doubt - I cannot merge or approve PRs in Mitogen. I have no permissions in the Github project, or PyPI.

**Comment by madrover** [comment]
This is an extremely sad state for such a useful project :(

Ideally it would be great to have this functionality built into the Ansible itself, which would likely imply better support. There's still an RFE open for that matter at https://github.com/ansible/ansible/issues/45839, but the latest outcome of the related discussions does not seem to be positive: https://meetbot.fedoraproject.org/ansible-meeting/2020-08-04/ansible_core_public_irc_meeting.2020-08-04-19.09.log.html I'd personally favor to lose some of the optimizations that Mitogen provides so we can see it built into Ansible...

It's worth noting that the discussions at https://github.com/ansible/ansible/issues/45839 are rather old. I don't know if with the advent of Ansible 2.10, which is less monolithic than previous version, can ease the way of contributing Mitogen to Ansible.

Alternatively, a project fork can make sense. That's one of the perks of opensource, isn't it? I personally think this is a sad approach but given the challenges contacting @dw, this option must be considered. For example, the project could be forked to something like https://github.com/ansible-community/ansible-mitogen and create a new library called ansible-mitogen. This is likely less complicated to do from a technical perspective than contributing to the Ansible core.

**Comment by moreati** [comment]
I've created https://github.com/mitoxec as a *potential* organisation to hold mitogen or similar. @dw and @s1113950, you've both been invited as co-owners. The name is also a placeholder, based in part on Github/PyPI/Twitter availability, and open to suggestions. Others I've considered include starpath, apomictic, subether, starjump, telexec, mitorun, farxec, genexec, mitovide - with the broad theme being faster than light communication, cell division, and action at a distance.

To reiterate - I don't want to fork Mitogen. It's a last option, other than abandoning the project.

**Comment by moreati** [comment]
@madrover I'm conscious that

a) Ansible may the main use of Mitogen, but it's not the only (possible) use. Mitogen can be used without Ansible
b) The dust is still settling on Ansible 2.10.0. I don't think Mitogen is well positioned to be under the Ansible banner (yet).

**Comment by apple4ever** [comment]
Sad to see this has fallen off. Operon doesn't seem to have taken off (the website hasn't been updated since 2019). Maybe is just not visibly active, but that really doesn't help us here. 

It seems like the options really are fork or let it die. I would like it to continue, but I would only do that if we forked. No reason to continue here if we will be blocked.

**Comment by moreati** [comment]
I'm publicising this RFC, to get wider input. So far
- https://mobile.twitter.com/moreati/status/1349453078823628800
- https://www.reddit.com/r/ansible/comments/kwp91k/rfc_sustainability_of_mitogen_an_ansible/?

**Comment by jrosser** [comment]
+1 for more folk being around in IRC. Aside from the technical issues discussed on this RFC, creating more of a community around Mitogen beyond GitHub issues/PR would be nice, as simple maybe as a designated 'office hour' in the IRC channel for folk to chat.

In terms of very few people making PRs - for those people who make the effort to produce a PR or an issue with a decent reproducer, the experience should to be good. At the end of the day you never know which user will turn out to be the next contributor, or even a future maintainer. Having active community, however you define that, is key to making that happen.

**Comment by jbscare** [comment]
We use Mitogen and love it and would love to see it continue, either as a plugin or as the standard Ansible engine.

We aren't able to offer help in supporting it at this time. :^(

**Comment by DWSR** [comment]
My 2 cents as an Ansible user both personally and professionally:

TL;DR it should be forked or be used as the inspiration for a new project so that 2.10 support can be added and so that this situation can be remedied.

* This project seems to contain a lot of "voodoo". I'm not sure how feasible unwinding that without the original author's help is.
* No matter what happens next, this project should be moved into its own GH org with at least 2 owners to prevent this from happening again.
* Now that Ansible 2.10 is out the door and the internals are in a better place, it might be a good time to take the spirit of Mitogen and turn it into something that plays nicer with Ansible itself. I don't know of any significant uses of Mitogen outside of Ansible, so I think cutting the scope of Mitogen to just be "a better Ansible Engine" feels reasonable and may eventually cause it to be pulled into mainline Ansible.
* The longer that this situation continues, the worse the "rubber band effect" will be when it is finally corrected. Ansible support lifecycles, the shuttering of travis-ci.org, and just general community atrophy all strongly suggest taking action ASAP.

**Comment by dw** [comment]
> I don't know of any significant uses of Mitogen

The core library has significant use outside of Ansible, and besides those folk have consistently demonstrated over a 3 year period an inability to even file bugs in this repo, what reason is there to believe they would they ever make suitable maintainers?

I'm reading this thread, and support and trust both Steven and Alex, please don't mistake being unresponsive to solicitations for free labour as a loss of interest.

**Comment by RonnyPfannschmidt** [comment]
i'd also like to be able to chime in with maintenance eventually, as mitogen is more cappable than execnet and i'd like to enable it to be a engine for pytest-xdist eventually (however in the near future my focus is elsewhere)

**Comment by moreati** [comment]
David, great to hear from you.

Thank you everyone who responded. It looks like we (David, Steven, and I) are going to discuss matters offline and work out a course of action.

**Comment by dw** [comment]
I've migrated the repo to TravisCI.com and rebuilt that master commit, but somehow it is still stuck. Meanwhile, whatever the migration was intended to fix should be working now

**Comment by dw** [comment]
Any idea what's up with the Mac build? Is it a 2.10 thing? Yowzer, 40 minutes :) https://dev.azure.com/dw-mitogen/Mitogen/_build/results?buildId=1143&view=logs&j=23e82cb1-eecb-5c20-63ca-90ff600fc34d

**Comment by moreati** [comment]
> Any idea what's up with the Mac build? Is it a 2.10 thing? Yowzer, 40 minutes :) https://dev.azure.com/dw-mitogen/Mitogen/_build/results?buildId=1143&view=logs&j=23e82cb1-eecb-5c20-63ca-90ff600fc34d

"Mac Vanilla_210_27" job is vanilla Ansible (no Mitogen, strategy=linear) https://github.com/dw/mitogen/blob/master/.ci/azure-pipelines.yml#L28-L32

**Comment by dw** [comment]
A little update for observers:

- [x] extra PyPI release managers added 🚀 -- informal agreement to have someone else run eye over the diffs before cutting a release, given the sensitivity of where Mitogen is used (my only major concern here)
- [x] Moved repo to an org -- Alex and Steven both can manage the repo integrations 🚀

**Comment by dw** [comment]
Marking fixed :)
